### DISCUSSION TITLE
New in v3.0.0 - ✨New and Improved Tools - Update Neovim from Chat Buffers

### DISCUSSION THREAD

**Discussion by olimorris** [olimorris/codecompanion.nvim#228] [Announcements]
Tools have been in the plugin since [mid-May](https://github.com/olimorris/codecompanion.nvim/commit/b36d9dc92dce83eced46ebe53279ee3fe9c8b90d). A feature I was proud of as I don't believe any Neovim plugin had developed any support for them at the time.

However, since their release I've felt they've been a somewhat underused feature of the plugin...both by myself and the user base. As of this v3.0.0 release, I'm adding in an `editor` tool which will make it much easier to update code in Neovim, from the chat buffer. Alongside the RAG and Code Runner tools, I'm hoping this will be a welcome addition.

Also, I've updated the [TOOLS.md](https://github.com/olimorris/codecompanion.nvim/blob/main/doc/TOOLS.md) file so you can start building your own.

## Improved Tools

> Note that in the videos below I'm confirming the tool via `cmp`...I removed this functionality before I launched v3.0.0.

**RAG**:

https://github.com/user-attachments/assets/dc9f0358-aaa1-4c58-8000-d2962257ffbd

**Editor** (adding a new function)

https://github.com/user-attachments/assets/01e028e1-bd2a-423c-8000-edbf79142095

**Editor** (replacing code)

https://github.com/user-attachments/assets/b5379313-22ea-4987-8106-67a0ba2545d7

## What's to come

- I'd hoped to have added a `/fetch` slash command but that will be later on in the release cycle.
- I need to add back support for `mini.diff`

**Comment by raine** [comment]
Hey again, not sure what would be a more appropriate place for this question, but...

Testing the `@editor` in chat, I'm noticing that contrary to the demos above, the assistant quite often does not directly change the code, but outputs the improvement as is in the chat, and then asks "Would you like me to update the script in the buffer for you?"

This is with default configuration.

Is there some way to make it more likely to:
- not output the improvement as is to save tokens
- just output the "tool xml" diff thing directly so that you don't have to waste time answering "Yes"

Thanks in advance, greatly appreciate the plugin!

**Reply by olimorris** [reply]
I've revised this in v3.5.0 which was released yesterday. I've tested mainly with Copilot and it's much more efficient:

![2024-09-24 08_28_20 - WezTerm@2x](https://github.com/user-attachments/assets/6345d286-77c6-4ff7-9b36-35b77fb939a9)

**Reply by olimorris** [reply]
Oh and I'm aiming to have more complex edits in the `editor` plugin very soon. So if the LLM decides it wants to delete some text and add text, the line numbers will automatically update (so no math for the LLM!).
